src/modeling.rs

```diff
@@ -113,13 +125,27 @@ impl Attention {
             k.reshape_like(&bsnh).unwrap().permute::<_, Tr12>()
         };
 
-        let v = {
+        let mut v = {
             let v_proj = self.v_proj.get_on(x.device());
             let v = x.matmul(v_proj.permute());
             v.reshape_like(&bsnh).unwrap().permute::<_, Tr12>()
         };
 
-        let (q, k) = self.rotary_embed.forward(q, k);
+        let (q, mut k) = self.rotary_embed.forward(
+            q,
+            k,
+            cache.as_ref().map(|c| c.key.shape().2).unwrap_or_default(),
+        );
+
+        if let Some(cache) = cache {
+            k = (cache.key, k).concat_along(Axis::<2>);
+            v = (cache.val, v).concat_along(Axis::<2>);
+        }
```

Comment:
We need a causal mask in this case, because previous k/vs wouldn't attend to the current tokens